### DISCUSSION TITLE
How to define many to many, one to many relationships with sqlalchemy without using foreign keys

### DISCUSSION THREAD

**Discussion by CaptionLittleBird** [sqlalchemy/sqlalchemy#11376] [Usage Questions]
I want to define the association between MySQL tables, but I don't want to sacrifice performance. How can I use SQLA to define logical foreign keys and avoid using physical foreign keys?

```
class UserAndScopes(Base):
    __tablename__ = 'user_and_scopes'
    id = mapped_column(Integer(), primary_key=True, autoincrement=True)
    user_id = mapped_column(ForeignKey('auth_user.id', ondelete="CASCADE"), primary_key=True)
    scopes_id = mapped_column(ForeignKey('auth_scopes.id'), primary_key=True)
```

**Comment by CaselIT** [comment]
Hi,

You car refer to https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#relationship-custom-foreign

**Reply by CaptionLittleBird** [reply]
I am very puzzled. My original intention was to use virtual foreign keys instead of physical foreign keys in MySQL. Can sqlalchemy do this? Can you provide an example of a model definition？

**Comment by CaptionLittleBird** [comment]
```
class Order(Base):
    __tablename__ = 'order'
    __table_args__ = {
        'comment': 'order table'
    }
    id = mapped_column(BigInteger(), primary_key=True, autoincrement=True)
    name: Mapped[Optional[normal_str]] = mapped_column(comment='name')
    org_id = mapped_column(Integer, comment='org', nullable=True)
    org: Mapped["Organization"] = relationship(back_populates='orders', primaryjoin='foreign(Organization.id) == Order.org_id')
```

```
class Organization(Base):
    __tablename__ = 'organization'
    __table_args__ = {
        'comment': 'org'
    }
    id = mapped_column(Integer(), primary_key=True, autoincrement=True)
    name: Mapped[Optional[normal_str]] = mapped_column(comment='name')
    password: Mapped[Optional[str]] = mapped_column(String(255), comment='pass')
    orders: Mapped[List["Order"]] = relationship(back_populates='org',primaryjoin='foreign(Order.org_id)==Organization.id') 
```
These are my two models，when I run the program, I keep reporting errors,

**sqlalchemy.exc.ArgumentError: Organization.orders and back-reference Order.org are both of the same direction <RelationshipDirection.ONETOMANY: 1>.  Did you mean to set remote_side on the many-to-one side ?**


How should I solve it?

**Reply by CaselIT** [reply]
You are using foreign wrong. Sadly the name are not great, but foreign indicates the column of the local table that should be treated as foring key.
Try usin `'remote(Organization.id) == foreign (Order.org_id)'` and `remote(Order.org_id)==foreign(Organization.id)'`
